src/Operators/integrals.jl

```diff
@@ -1,315 +1,371 @@
+`output`. The `input` can be either a `Field` or an `AbstractBroadcasted` that
+performs pointwise or columnwise operations on `Field`s. Each reduced value is
+computed by iteratively applying `f` to the values in `input`, starting from the
+bottom of each column and moving upward, and the result of the final iteration
+is passed to the `transform` function before being stored in `output`. If `init`
+is specified, it is used as the initial value of the iteration; otherwise, the
+value at the bottom of each column in `input` is used as the initial value.
+    
+With `first_level` and `last_level` denoting the indices of the boundary levels
+of `input`, the reduction in each column can be summarized as follows:
+  - If `init` is unspecified,
+    ```
+    reduced_value = input[first_level]
+    for level in (first_level + 1):last_level
+        reduced_value = f(reduced_value, input[level - half])
+    end
+    output[] = transform(reduced_value)
+    ```
+  - If `init` is specified,
+    ```
+    reduced_value = init
+    for level in first_level:last_level
+        reduced_value = f(reduced_value, input[level])
+    end
+    output[] = transform(reduced_value)
+    ```
+"""
+function column_reduce!(
+    f::F,
+    output::Fields.Field,
+    input::Union{Fields.Field, PointwiseOrColumnwiseBroadcasted};
+    init = UnspecifiedInit(),
+    transform::T = identity,
+) where {F, T}
+    device = ClimaComms.device(output)
+    column_reduce_device!(device, f, transform, output, input, init)
+end
 
-column_integral_indefinite!(
-    f::Function,
+column_reduce_device!(
     ::ClimaComms.AbstractCPUDevice,
-    ᶠ∫field,
-    args...,
-) = column_integral_indefinite_cpu!(f, ᶠ∫field, args...)
-
-column_integral_indefinite!(
-    f::Function,
+    f::F,
+    transform::T,
+    output::Fields.SpectralElementField,
+    input,
+    init,
+) where {F, T} =
+    Fields.bycolumn(axes(output)) do colidx
+        single_column_reduce!(
+            f,
+            transform,
+            output[colidx],
+            extract_column(input, colidx),
+            init,
+        )
+    end
+column_reduce_device!(
     ::ClimaComms.AbstractCPUDevice,
-    ᶠ∫field::Fields.FaceExtrudedFiniteDifferenceField,
-    args...,
-) =
-    Fields.bycolumn(axes(ᶠ∫field)) do colidx
-        column_integral_indefinite_cpu!(f, ᶠ∫field[colidx], args...)
+    f::F,
+    transform::T,
+    output::Fields.PointField,
+    input,
+    init,
+) where {F, T} = single_column_reduce!(f, transform, output, input, init)
+
+function single_column_reduce!(
+    f::F,
+    transform::T,
+    output,
+    input,
+    init,
+) where {F, T}
+    first_level = left_boundary_idx(axes(input))
+    last_level = right_boundary_idx(axes(input))
+    @inbounds if init == UnspecifiedInit()
+        reduced_value = column_value_at_level(input, first_level)
+        level_range = (first_level + 1):last_level
+    else
+        reduced_value = init
```

Comment:
```suggestion
        reduced_value::typeof(init) = init
```
It might turn out to be a good idea to this assertion here, so that the compiler doesn't need to infer through the entire `level_range` (since we're not doing `reduce_value .= ...`)

Reply:
(and same above)

Reply:
There aren't any type instabilities in the GPU tests at the moment, so it doesn't look like the additional typecast makes a difference here.